pkg/treeoperator/controller/virtualcluster_join_controller.go

```diff
@@ -1,16 +1,452 @@
 package controller
 
 import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	extensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/kosmos.io/kosmos/pkg/apis/kosmos/v1alpha1"
+	"github.com/kosmos.io/kosmos/pkg/cert"
+	clusterManager "github.com/kosmos.io/kosmos/pkg/clustertree/cluster-manager"
+	"github.com/kosmos.io/kosmos/pkg/generated/clientset/versioned"
+	"github.com/kosmos.io/kosmos/pkg/kosmosctl/manifest"
+	"github.com/kosmos.io/kosmos/pkg/kosmosctl/util"
+	kosmosctl "github.com/kosmos.io/kosmos/pkg/kosmosctl/util"
+	"github.com/kosmos.io/kosmos/pkg/treeoperator/constants"
+	treemanifest "github.com/kosmos.io/kosmos/pkg/treeoperator/manifest"
+	"github.com/kosmos.io/kosmos/pkg/utils"
+	"github.com/kosmos.io/kosmos/pkg/version"
 )
 
 type VirtualClusterJoinController struct {
 	client.Client
-	EventRecorder record.EventRecorder
+	EventRecorder    record.EventRecorder
+	KubeconfigPath   string
+	KubeconfigStream []byte
 }
 
-func (c *VirtualClusterJoinController) SetupWithManager(mgr manager.Manager) error {
+func (c *VirtualClusterJoinController) RemoveClusterFinalizer(cluster *v1alpha1.Cluster, kosmosClient versioned.Interface) error {
+	for _, finalizer := range []string{utils.ClusterStartControllerFinalizer, clusterManager.ControllerFinalizerName} {
+		if controllerutil.ContainsFinalizer(cluster, finalizer) {
+			controllerutil.RemoveFinalizer(cluster, finalizer)
+		}
+	}
+	klog.Infof("remove finalizer for cluster %s", cluster.Name)
+
+	_, err := kosmosClient.KosmosV1alpha1().Clusters().Update(context.TODO(), cluster, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Errorf("cluster %s failed remove finalizer: %v", cluster.Name, err)
+		return err
+	}
+	klog.Infof("update cluster after remove finalizer for cluster %s", cluster.Name)
+	return nil
+}
+
+func (c *VirtualClusterJoinController) UninstallClustertree(ctx context.Context, request reconcile.Request, vc *v1alpha1.VirtualCluster) error {
```

Comment:
The first character T of tree needs to be capitalized ｜ ClusterTree

Reply:
fix it